syntax = "proto3";

package raft;

service Raft {
    rpc RequestVote(VoteRequest) returns (VoteResponse);
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
    rpc ClientRequest(ClientRequestMsg) returns (ClientResponse);
    rpc NetworkPartition(PartitionRequest) returns (PartitionResponse);
}

message VoteRequest {
    int32 term = 1;
    int32 candidate_id = 2;
    int32 last_log_index = 3;
    int32 last_log_term = 4;
}

message VoteResponse {
    int32 term = 1;
    bool vote_granted = 2;
}

message LogEntry {
    int32 term = 1;
    int32 index = 2;
    string command = 3;
}

message AppendEntriesRequest {
    int32 term = 1;
    int32 leader_id = 2;
    int32 prev_log_index = 3;
    int32 prev_log_term = 4;
    repeated LogEntry entries = 5;
    int32 leader_commit = 6;
}

message AppendEntriesResponse {
    int32 term = 1;
    bool success = 2;
}

message ClientRequestMsg {
    string command = 1;
}

message ClientResponse {
    bool success = 1;
    string message = 2;
    int32 leader_id = 3;
}

message PartitionRequest {
    repeated string blocked_addresses = 1;
}

message PartitionResponse {
    bool success = 1;
}